// @aws-amplify/ui-components/dist/cjs/amplify-confirm-sign-in_7.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

const index = require('./index-6d44143a.js');
const core = require('@aws-amplify/core');
const authTypes = require('./auth-types-cd1f71d2.js');
const Auth = require('@aws-amplify/auth');
const Auth__default = _interopDefault(Auth);
const Translations = require('./Translations-5f4bb72c.js');
const constants = require('./constants-1335fef8.js');
const helpers = require('./helpers-0d7ea2c5.js');
const authHelpers = require('./auth-helpers-ed56b901.js');

const AmplifyConfirmSignIn = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /** Fires when confirm sign in form is submitted */
        this.handleSubmit = event => this.confirm(event);
        /** Used for header text in confirm sign in component */
        this.headerText = Translations.Translations.CONFIRM_SMS_CODE;
        /** Used for the submit button text in confirm sign in component */
        this.submitButtonText = Translations.Translations.CONFIRM;
        /** Auth state change handler for this component */
        this.handleAuthStateChange = helpers.dispatchAuthStateChangeEvent;
        /** Default form field */
        this.defaultFormFields = [
            {
                type: 'code',
                required: true,
                handleInputChange: event => this.handleCodeChange(event),
            },
        ];
        /**
         * Form fields allows you to utilize our pre-built components such as username field, code field, password field, email field, etc.
         * by passing an array of strings that you would like the order of the form to be in. If you need more customization, such as changing
         * text for a label or adjust a placeholder, you can follow the structure below in order to do just that.
         * ```
         * [
         *  {
         *    type: string,
         *    label: string,
         *    placeholder: string,
         *    hint: string | Functional Component | null,
         *    required: boolean
         *  }
         * ]
         * ```
         */
        this.formFields = this.defaultFormFields;
        /** The MFA option to confirm with */
        this.mfaOption = authTypes.MfaOption.SMS;
        /* Whether or not the confirm-sign-in component is loading */
        this.loading = false;
    }
    componentWillLoad() {
        this.setup();
    }
    userHandler() {
        this.setup();
    }
    setup() {
        if (this.user &&
            this.user['challengeName'] === authTypes.ChallengeName.SoftwareTokenMFA) {
            this.mfaOption = authTypes.MfaOption.TOTP;
            // If header text is using default use TOTP string
            if (this.headerText === Translations.Translations.CONFIRM_SMS_CODE) {
                this.headerText = Translations.Translations.CONFIRM_TOTP_CODE;
            }
        }
        this.constructedFormFieldOptions = this.constructFormFieldOptions(this.formFields);
    }
    handleCodeChange(event) {
        this.code = event.target.value;
    }
    async confirm(event) {
        if (event) {
            event.preventDefault();
        }
        const mfaType = this.user['challengeName'] === authTypes.ChallengeName.SoftwareTokenMFA
            ? authTypes.ChallengeName.SoftwareTokenMFA
            : null;
        if (!Auth.Auth || typeof Auth.Auth.confirmSignIn !== 'function') {
            throw new Error(constants.NO_AUTH_MODULE_FOUND);
        }
        this.loading = true;
        try {
            await Auth.Auth.confirmSignIn(this.user, this.code, mfaType);
            await authHelpers.checkContact(this.user, this.handleAuthStateChange);
        }
        catch (error) {
            helpers.dispatchToastHubEvent(error);
        }
        finally {
            this.loading = false;
        }
    }
    constructFormFieldOptions(formFields) {
        const content = [];
        if (formFields === undefined)
            return undefined;
        if (formFields.length <= 0)
            return this.defaultFormFields;
        formFields.forEach((formField) => {
            if (typeof formField === 'string' || formField.type !== 'code') {
                // This is either a `string`, and/or a custom field that isn't `code`. Pass this directly.
                content.push(formField);
            }
            else {
                // This is a code input field. Attach input handler.
                content.push(Object.assign({ handleInputChange: event => this.handleCodeChange(event) }, formField));
            }
        });
        return content;
    }
    render() {
        return (index.h(index.Host, null, index.h("amplify-form-section", { headerText: core.I18n.get(this.headerText), handleSubmit: this.handleSubmit, submitButtonText: core.I18n.get(this.submitButtonText), loading: this.loading, secondaryFooterContent: index.h("span", null, index.h("amplify-button", { variant: "anchor", onClick: () => this.handleAuthStateChange(authTypes.AuthState.SignIn) }, core.I18n.get(Translations.Translations.BACK_TO_SIGN_IN))) }, index.h("amplify-auth-fields", { formFields: this.constructedFormFieldOptions }))));
    }
    static get watchers() { return {
        "user": ["userHandler"]
    }; }
};

const AmplifyConfirmSignUp = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /** Fires when sign up form is submitted */
        this.handleSubmit = event => this.confirmSignUp(event);
        /** Used for header text in confirm sign up component */
        this.headerText = Translations.Translations.CONFIRM_SIGN_UP_HEADER_TEXT;
        /** Used for the submit button text in confirm sign up component */
        this.submitButtonText = Translations.Translations.CONFIRM_SIGN_UP_SUBMIT_BUTTON_TEXT;
        /**
         * Form fields allows you to utilize our pre-built components such as username field, code field, password field, email field, etc.
         * by passing an array of strings that you would like the order of the form to be in. If you need more customization, such as changing
         * text for a label or adjust a placeholder, you can follow the structure below in order to do just that.
         * ```
         * [
         *  {
         *    type: string,
         *    label: string,
         *    placeholder: string,
         *    hint: string | Functional Component | null,
         *    required: boolean
         *  }
         * ]
         * ```
         */
        this.formFields = [];
        /** Auth state change handler for this components
         * e.g. SignIn -> 'Create Account' link -> SignUp
         */
        this.handleAuthStateChange = helpers.dispatchAuthStateChangeEvent;
        /** Username Alias is used to setup authentication with `username`, `email` or `phone_number`  */
        this.usernameAlias = 'username';
        this.loading = false;
        this.newFormFields = [];
        this.phoneNumber = {
            countryDialCodeValue: constants.COUNTRY_DIAL_CODE_DEFAULT,
            phoneNumberValue: null,
        };
    }
    componentWillLoad() {
        this.setup();
    }
    formFieldsHandler() {
        this.buildFormFields();
    }
    userHandler() {
        this.setup();
    }
    setup() {
        // TODO: Use optional chaining instead
        this.userInput = this.user && this.user.username;
        this._signUpAttrs = this.user && this.user.signUpAttrs;
        helpers.checkUsernameAlias(this.usernameAlias);
        this.buildFormFields();
    }
    buildDefaultFormFields() {
        this.newFormFields = [
            {
                type: `${this.usernameAlias}`,
                required: true,
                handleInputChange: this.handleFormFieldInputChange(`${this.usernameAlias}`),
                value: this.userInput,
                disabled: this.userInput ? true : false,
                inputProps: { autocomplete: 'username' },
            },
            {
                type: 'code',
                label: core.I18n.get(Translations.Translations.CONFIRM_SIGN_UP_CODE_LABEL),
                placeholder: core.I18n.get(Translations.Translations.CONFIRM_SIGN_UP_CODE_PLACEHOLDER),
                required: true,
                hint: (index.h("div", null, core.I18n.get(Translations.Translations.CONFIRM_SIGN_UP_LOST_CODE), ' ', index.h("amplify-button", { variant: "anchor", onClick: () => this.resendConfirmCode() }, core.I18n.get(Translations.Translations.CONFIRM_SIGN_UP_RESEND_CODE)))),
                handleInputChange: this.handleFormFieldInputChange('code'),
            },
        ];
    }
    buildFormFields() {
        if (this.formFields.length === 0) {
            this.buildDefaultFormFields();
        }
        else {
            const newFields = [];
            this.formFields.forEach(field => {
                const newField = Object.assign({}, field);
                if (newField.type === 'code') {
                    newField['hint'] = helpers.isHintValid(newField) ? (index.h("div", null, core.I18n.get(Translations.Translations.CONFIRM_SIGN_UP_LOST_CODE), ' ', index.h("amplify-button", { variant: "anchor", onClick: () => this.resendConfirmCode() }, core.I18n.get(Translations.Translations.CONFIRM_SIGN_UP_RESEND_CODE)))) : (newField['hint']);
                }
                newField['handleInputChange'] = event => this.handleFormFieldInputWithCallback(event, field);
                newFields.push(newField);
            });
            this.newFormFields = newFields;
        }
    }
    handleFormFieldInputChange(fieldType) {
        switch (fieldType) {
            case 'username':
            case 'email':
                return event => (this.userInput = event.target.value);
            case 'phone_number':
                return event => helpers.handlePhoneNumberChange(event, this.phoneNumber);
            case 'code':
                return event => (this.code = event.target.value);
            default:
                return;
        }
    }
    setFieldValue(field) {
        switch (field.type) {
            case 'username':
            case 'email':
                if (field.value === undefined) {
                    this.userInput = '';
                }
                else {
                    this.userInput = field.value;
                }
                break;
            case 'phone_number':
                if (field.dialCode) {
                    this.phoneNumber.countryDialCodeValue = field.dialCode;
                }
                this.phoneNumber.phoneNumberValue = field.value;
                break;
        }
    }
    handleFormFieldInputWithCallback(event, field) {
        const fnToCall = field['handleInputChange']
            ? field['handleInputChange']
            : (event, cb) => {
                cb(event);
            };
        const callback = this.handleFormFieldInputChange(field.type);
        fnToCall(event, callback.bind(this));
    }
    /**
     * Returns the username of the user to confirm. If a valid `user.username` has been passed as a prop, we return that.
     * Otherwise, we return the `userInput` on the form.
     */
    getUsername() {
        if (this.user && this.user.username)
            return this.user.username;
        switch (this.usernameAlias) {
            case 'username':
            case 'email':
                return this.userInput;
            case 'phone_number':
                return helpers.composePhoneNumberInput(this.phoneNumber);
        }
    }
    async resendConfirmCode() {
        if (event) {
            event.preventDefault();
        }
        if (!Auth.Auth || typeof Auth.Auth.resendSignUp !== 'function') {
            throw new Error(constants.NO_AUTH_MODULE_FOUND);
        }
        try {
            const username = this.getUsername();
            if (!username)
                throw new Error(Translations.Translations.EMPTY_USERNAME);
            await Auth.Auth.resendSignUp(username.trim());
        }
        catch (error) {
            helpers.dispatchToastHubEvent(error);
        }
    }
    // TODO: Add validation
    // TODO: Prefix
    async confirmSignUp(event) {
        if (event) {
            event.preventDefault();
        }
        if (!Auth.Auth || typeof Auth.Auth.confirmSignUp !== 'function') {
            throw new Error(constants.NO_AUTH_MODULE_FOUND);
        }
        this.loading = true;
        try {
            let username = this.getUsername();
            if (!username)
                throw new Error(Translations.Translations.EMPTY_USERNAME);
            username = username.trim();
            const confirmSignUpResult = await Auth.Auth.confirmSignUp(username, this.code);
            if (!confirmSignUpResult) {
                throw new Error(core.I18n.get(Translations.Translations.CONFIRM_SIGN_UP_FAILED));
            }
            if (this._signUpAttrs &&
                this._signUpAttrs.password &&
                this._signUpAttrs.password !== '') {
                // Auto sign in user if password is available from previous workflow
                await authHelpers.handleSignIn(username, this._signUpAttrs.password, this.handleAuthStateChange);
            }
            else {
                this.handleAuthStateChange(authTypes.AuthState.SignIn);
            }
        }
        catch (error) {
            helpers.dispatchToastHubEvent(error);
        }
        finally {
            this.loading = false;
        }
    }
    render() {
        return (index.h(index.Host, null, index.h("amplify-form-section", { headerText: core.I18n.get(this.headerText), submitButtonText: core.I18n.get(this.submitButtonText), handleSubmit: this.handleSubmit, loading: this.loading, secondaryFooterContent: index.h("div", null, index.h("span", null, index.h("amplify-button", { variant: "anchor", onClick: () => this.handleAuthStateChange(authTypes.AuthState.SignIn) }, core.I18n.get(Translations.Translations.BACK_TO_SIGN_IN)))) }, index.h("amplify-auth-fields", { formFields: this.newFormFields }))));
    }
    static get watchers() { return {
        "formFields": ["formFieldsHandler"],
        "user": ["userHandler"]
    }; }
};

const logger = new core.Logger('ForgotPassword');
const AmplifyForgotPassword = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /** The header text of the forgot password section */
        this.headerText = Translations.Translations.RESET_YOUR_PASSWORD;
        /** The text displayed inside of the send code button for the form */
        this.sendButtonText = Translations.Translations.SEND_CODE;
        /** The text displayed inside of the submit button for the form */
        this.submitButtonText = Translations.Translations.SUBMIT;
        /** The form fields displayed inside of the forgot password form */
        this.formFields = [];
        /** The function called when making a request to reset password */
        this.handleSend = event => this.send(event);
        /** The function called when submitting a new password */
        this.handleSubmit = event => this.submit(event);
        /** Auth state change handler for this component */
        this.handleAuthStateChange = helpers.dispatchAuthStateChangeEvent;
        /** Username Alias is used to setup authentication with `username`, `email` or `phone_number`  */
        this.usernameAlias = 'username';
        this.delivery = null;
        this.loading = false;
        this.phoneNumber = {
            countryDialCodeValue: constants.COUNTRY_DIAL_CODE_DEFAULT,
            phoneNumberValue: null,
        };
        this.newFormFields = [];
        this.forgotPasswordAttrs = {
            userInput: '',
            password: '',
            code: '',
        };
    }
    componentWillLoad() {
        helpers.checkUsernameAlias(this.usernameAlias);
        this.buildFormFields();
    }
    formFieldsHandler() {
        this.buildFormFields();
    }
    buildFormFields() {
        if (this.formFields.length === 0) {
            this.buildDefaultFormFields();
        }
        else {
            const newFields = [];
            this.formFields.forEach(field => {
                const newField = Object.assign({}, field);
                newField['handleInputChange'] = event => this.handleFormFieldInputWithCallback(event, field);
                newFields.push(newField);
            });
            this.newFormFields = newFields;
        }
    }
    buildDefaultFormFields() {
        switch (this.usernameAlias) {
            case 'email':
                this.newFormFields = [
                    {
                        type: 'email',
                        required: true,
                        handleInputChange: this.handleFormFieldInputChange('email'),
                        inputProps: {
                            'data-test': 'forgot-password-email-input',
                        },
                    },
                ];
                break;
            case 'phone_number':
                this.newFormFields = [
                    {
                        type: 'phone_number',
                        required: true,
                        handleInputChange: this.handleFormFieldInputChange('phone_number'),
                        inputProps: {
                            'data-test': 'forgot-password-phone-number-input',
                        },
                    },
                ];
                break;
            case 'username':
            default:
                this.newFormFields = [
                    {
                        type: 'username',
                        required: true,
                        handleInputChange: this.handleFormFieldInputChange('username'),
                        inputProps: {
                            'data-test': 'forgot-password-username-input',
                        },
                    },
                ];
                break;
        }
    }
    handleFormFieldInputChange(fieldType) {
        switch (fieldType) {
            case 'username':
            case 'email':
                return event => (this.forgotPasswordAttrs.userInput = event.target.value);
            case 'phone_number':
                return event => helpers.handlePhoneNumberChange(event, this.phoneNumber);
            case 'password':
            case 'code':
                return event => (this.forgotPasswordAttrs[fieldType] = event.target.value);
            default:
                return;
        }
    }
    setFieldValue(field, formAttributes) {
        switch (field.type) {
            case 'username':
            case 'email':
                if (field.value === undefined) {
                    formAttributes.userInput = '';
                }
                else {
                    formAttributes.userInput = field.value;
                }
                break;
            case 'phone_number':
                if (field.dialCode) {
                    this.phoneNumber.countryDialCodeValue = field.dialCode;
                }
                this.phoneNumber.phoneNumberValue = field.value;
                break;
            case 'password':
            case 'code':
                if (field.value === undefined) {
                    formAttributes[field.type] = '';
                }
                else {
                    formAttributes[field.type] = field.value;
                }
                break;
        }
    }
    handleFormFieldInputWithCallback(event, field) {
        const fnToCall = field['handleInputChange']
            ? field['handleInputChange']
            : (event, cb) => {
                cb(event);
            };
        const callback = this.handleFormFieldInputChange(field.type);
        fnToCall(event, callback.bind(this));
    }
    async send(event) {
        if (event) {
            event.preventDefault();
        }
        if (!Auth.Auth || typeof Auth.Auth.forgotPassword !== 'function') {
            throw new Error(constants.NO_AUTH_MODULE_FOUND);
        }
        this.loading = true;
        switch (this.usernameAlias) {
            case 'phone_number':
                try {
                    this.forgotPasswordAttrs.userInput = helpers.composePhoneNumberInput(this.phoneNumber);
                }
                catch (error) {
                    helpers.dispatchToastHubEvent(error);
                }
                break;
        }
        try {
            const data = await Auth.Auth.forgotPassword(this.forgotPasswordAttrs.userInput.trim());
            logger.debug(data);
            this.newFormFields = [
                {
                    type: 'code',
                    required: true,
                    handleInputChange: this.handleFormFieldInputChange('code'),
                    inputProps: {
                        'data-test': 'forgot-password-code-input',
                    },
                },
                {
                    type: 'password',
                    required: true,
                    handleInputChange: this.handleFormFieldInputChange('password'),
                    label: core.I18n.get(Translations.Translations.NEW_PASSWORD_LABEL),
                    placeholder: core.I18n.get(Translations.Translations.NEW_PASSWORD_PLACEHOLDER),
                },
            ];
            this.delivery = data.CodeDeliveryDetails;
        }
        catch (error) {
            helpers.dispatchToastHubEvent(error);
        }
        finally {
            this.loading = false;
        }
    }
    async submit(event) {
        if (event) {
            event.preventDefault();
        }
        if (!Auth.Auth || typeof Auth.Auth.forgotPasswordSubmit !== 'function') {
            throw new Error(constants.NO_AUTH_MODULE_FOUND);
        }
        this.loading = true;
        try {
            const { userInput, code, password } = this.forgotPasswordAttrs;
            const data = await Auth.Auth.forgotPasswordSubmit(userInput.trim(), code, password);
            logger.debug(data);
            this.handleAuthStateChange(authTypes.AuthState.SignIn);
            this.delivery = null;
        }
        catch (error) {
            helpers.dispatchToastHubEvent(error);
        }
        finally {
            this.loading = false;
        }
    }
    render() {
        const submitFn = this.delivery
            ? event => this.handleSubmit(event)
            : event => this.handleSend(event);
        const submitButtonText = this.delivery
            ? this.submitButtonText
            : this.sendButtonText;
        return (index.h(index.Host, null, index.h("amplify-form-section", { headerText: core.I18n.get(this.headerText), handleSubmit: submitFn, loading: this.loading, secondaryFooterContent: index.h("amplify-button", { variant: "anchor", onClick: () => this.handleAuthStateChange(authTypes.AuthState.SignIn), "data-test": "forgot-password-back-to-sign-in-link" }, core.I18n.get(Translations.Translations.BACK_TO_SIGN_IN)), testDataPrefix: 'forgot-password', submitButtonText: core.I18n.get(submitButtonText) }, index.h("amplify-auth-fields", { formFields: this.newFormFields }))));
    }
    static get watchers() { return {
        "formFields": ["formFieldsHandler"]
    }; }
};

const logger$1 = new core.ConsoleLogger('amplify-require-new-password');
const AmplifyRequireNewPassword = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /** The header text of the forgot password section */
        this.headerText = Translations.Translations.CHANGE_PASSWORD;
        /** The text displayed inside of the submit button for the form */
        this.submitButtonText = Translations.Translations.CHANGE_PASSWORD_ACTION;
        /** The function called when submitting a new password */
        this.handleSubmit = event => this.completeNewPassword(event);
        /** Auth state change handler for this component */
        this.handleAuthStateChange = helpers.dispatchAuthStateChangeEvent;
        /** The form fields displayed inside of the forgot password form */
        this.formFields = [
            {
                type: authTypes.AuthFormField.Password,
                required: true,
                handleInputChange: event => this.handlePasswordChange(event),
                label: core.I18n.get(Translations.Translations.NEW_PASSWORD_LABEL),
                placeholder: core.I18n.get(Translations.Translations.NEW_PASSWORD_PLACEHOLDER),
                inputProps: {
                    'data-test': 'require-new-password-password-input',
                },
            },
        ];
        this.loading = false;
        this.requiredAttributes = {};
        this.newFormFields = this.formFields;
        this.phoneNumber = {
            countryDialCodeValue: constants.COUNTRY_DIAL_CODE_DEFAULT,
            phoneNumberValue: ' ',
        };
    }
    userHandler() {
        this.setCurrentUser();
    }
    handleRequiredAttributeInputChange(attribute, event) {
        if (attribute === 'phone_number') {
            this.formatPhoneNumber(event);
        }
        else {
            this.requiredAttributes[attribute] = event.target.value;
        }
    }
    async setCurrentUser() {
        if (!this.user) {
            // Check for authenticated user
            try {
                const user = await Auth.Auth.currentAuthenticatedUser();
                if (user)
                    this.currentUser = user;
            }
            catch (error) {
                helpers.dispatchToastHubEvent(error);
            }
        }
        else {
            this.currentUser = this.user;
        }
        if (this.currentUser &&
            this.currentUser.challengeParam &&
            this.currentUser.challengeParam.requiredAttributes) {
            const userRequiredAttributes = this.currentUser.challengeParam
                .requiredAttributes;
            const requiredAttributesMap = helpers.getRequiredAttributesMap();
            userRequiredAttributes.forEach((attribute) => {
                const formField = {
                    type: attribute,
                    required: true,
                    label: requiredAttributesMap[attribute].label,
                    placeholder: requiredAttributesMap[attribute].placeholder,
                    handleInputChange: event => this.handleRequiredAttributeInputChange(attribute, event),
                    inputProps: {
                        'data-test': `require-new-password-${attribute}-input`,
                    },
                };
                this.newFormFields = [...this.newFormFields, formField];
            });
        }
    }
    componentWillLoad() {
        return this.setCurrentUser();
    }
    handlePasswordChange(event) {
        this.password = event.target.value;
    }
    formatPhoneNumber(event) {
        helpers.handlePhoneNumberChange(event, this.phoneNumber);
        /**
         * composePhoneNumberInput will throw an error if the phoneNumberValue it receives
         * is empty. Adding a check here to try and make sure that doesn't happen...but will
         * wrap it in a try/catch block just in case as well
         */
        try {
            if (event.target.name === constants.PHONE_SUFFIX &&
                this.phoneNumber.phoneNumberValue) {
                const composedInput = helpers.composePhoneNumberInput(this.phoneNumber);
                this.requiredAttributes['phone_number'] = composedInput;
            }
        }
        catch (err) {
            logger$1.error(`error in phone number field - ${err}`);
        }
    }
    async completeNewPassword(event) {
        if (event) {
            event.preventDefault();
        }
        if (!Auth.Auth || typeof Auth.Auth.completeNewPassword !== 'function') {
            throw new Error(constants.NO_AUTH_MODULE_FOUND);
        }
        this.loading = true;
        try {
            const user = await Auth.Auth.completeNewPassword(this.currentUser, this.password, this.requiredAttributes);
            logger$1.debug('complete new password', user);
            switch (user.challengeName) {
                case authTypes.ChallengeName.SMSMFA:
                    this.handleAuthStateChange(authTypes.AuthState.ConfirmSignIn, user);
                    break;
                case authTypes.ChallengeName.MFASetup:
                    logger$1.debug('TOTP setup', user.challengeParam);
                    this.handleAuthStateChange(authTypes.AuthState.TOTPSetup, user);
                    break;
                default:
                    await authHelpers.checkContact(user, this.handleAuthStateChange);
            }
        }
        catch (error) {
            helpers.dispatchToastHubEvent(error);
        }
        finally {
            this.loading = false;
        }
    }
    render() {
        return (index.h(index.Host, null, index.h("amplify-form-section", { headerText: core.I18n.get(this.headerText), handleSubmit: this.handleSubmit, loading: this.loading, secondaryFooterContent: index.h("amplify-button", { variant: "anchor", onClick: () => this.handleAuthStateChange(authTypes.AuthState.SignIn) }, core.I18n.get(Translations.Translations.BACK_TO_SIGN_IN)), submitButtonText: core.I18n.get(this.submitButtonText) }, index.h("amplify-auth-fields", { formFields: this.newFormFields }))));
    }
    static get watchers() { return {
        "user": ["userHandler"]
    }; }
};

const amplifySignInCss = ":host{--footer-size:var(--amplify-text-sm);--footer-color:var(--amplify-grey);--footer-font-family:var(--amplify-font-family);--font-weight:var(--amplify-font-weight)}.sign-in-form-footer{font-family:var(--footer-font-family);font-size:var(--footer-size);color:var(--footer-color);font-weight:--font-weight;display:-ms-flexbox;display:flex;-ms-flex-direction:column-reverse;flex-direction:column-reverse;-ms-flex-align:center;align-items:center;-ms-flex-pack:distribute;justify-content:space-around}.sign-in-form-footer amplify-button{margin-bottom:0.625rem}@media (min-width: 672px){.sign-in-form-footer{display:-ms-flexbox;display:flex;-ms-flex-direction:row;flex-direction:row;-ms-flex-pack:justify;justify-content:space-between;-ms-flex-align:baseline;align-items:baseline}.sign-in-form-footer amplify-button{margin-bottom:0}}.sign-in-form-footer *+*{margin-bottom:15px}.full-width-footer-content{width:100%}";

const AmplifySignIn = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /** Fires when sign in form is submitted */
        this.handleSubmit = event => this.signIn(event);
        /** Used for header text in sign in component */
        this.headerText = Translations.Translations.SIGN_IN_HEADER_TEXT;
        /** Used for the submit button text in sign in component */
        this.submitButtonText = Translations.Translations.SIGN_IN_ACTION;
        /** Auth state change handler for this component */
        // prettier-ignore
        this.handleAuthStateChange = helpers.dispatchAuthStateChangeEvent;
        /** Username Alias is used to setup authentication with `username`, `email` or `phone_number`  */
        this.usernameAlias = 'username';
        /**
         * Form fields allows you to utilize our pre-built components such as username field, code field, password field, email field, etc.
         * by passing an array of strings that you would like the order of the form to be in. If you need more customization, such as changing
         * text for a label or adjust a placeholder, you can follow the structure below in order to do just that.
         * ```
         * [
         *  {
         *    type: string,
         *    label: string,
         *    placeholder: string,
         *    hint: string | Functional Component | null,
         *    required: boolean
         *  }
         * ]
         * ```
         */
        this.formFields = [];
        /** Hides the sign up link */
        this.hideSignUp = false;
        this.newFormFields = [];
        /* Whether or not the sign-in component is loading */
        this.loading = false;
        this.phoneNumber = {
            countryDialCodeValue: constants.COUNTRY_DIAL_CODE_DEFAULT,
            phoneNumberValue: null,
        };
        this.signInAttributes = {
            userInput: '',
            password: '',
        };
    }
    componentWillLoad() {
        helpers.checkUsernameAlias(this.usernameAlias);
        this.buildFormFields();
    }
    formFieldsHandler() {
        this.buildFormFields();
    }
    handleFormFieldInputChange(fieldType) {
        switch (fieldType) {
            case 'username':
            case 'email':
                return event => (this.signInAttributes.userInput = event.target.value);
            case 'phone_number':
                return event => helpers.handlePhoneNumberChange(event, this.phoneNumber);
            case 'password':
                return event => (this.signInAttributes.password = event.target.value);
            default:
                return () => { };
        }
    }
    handleFormFieldInputWithCallback(event, field) {
        const fnToCall = field['handleInputChange']
            ? field['handleInputChange']
            : (event, cb) => {
                cb(event);
            };
        const callback = this.handleFormFieldInputChange(field.type);
        fnToCall(event, callback.bind(this));
    }
    async signIn(event) {
        // avoid submitting the form
        if (event) {
            event.preventDefault();
        }
        this.loading = true;
        switch (this.usernameAlias) {
            case 'phone_number':
                try {
                    this.signInAttributes.userInput = helpers.composePhoneNumberInput(this.phoneNumber);
                }
                catch (error) {
                    helpers.dispatchToastHubEvent(error);
                }
        }
        const username = this.signInAttributes.userInput.trim();
        await authHelpers.handleSignIn(username, this.signInAttributes.password, this.handleAuthStateChange, this.usernameAlias);
        this.loading = false;
    }
    buildDefaultFormFields() {
        const formFieldInputs = [];
        switch (this.usernameAlias) {
            case 'email':
                formFieldInputs.push({
                    type: 'email',
                    required: true,
                    handleInputChange: this.handleFormFieldInputChange('email'),
                    inputProps: {
                        'data-test': 'sign-in-email-input',
                        autocomplete: 'username',
                    },
                });
                break;
            case 'phone_number':
                formFieldInputs.push({
                    type: 'phone_number',
                    required: true,
                    handleInputChange: this.handleFormFieldInputChange('phone_number'),
                    inputProps: {
                        'data-test': 'sign-in-phone-number-input',
                        autocomplete: 'username',
                    },
                });
                break;
            case 'username':
            default:
                formFieldInputs.push({
                    type: 'username',
                    required: true,
                    handleInputChange: this.handleFormFieldInputChange('username'),
                    inputProps: {
                        'data-test': 'sign-in-username-input',
                        autocomplete: 'username',
                    },
                });
                break;
        }
        formFieldInputs.push({
            type: 'password',
            hint: (index.h("div", null, core.I18n.get(Translations.Translations.FORGOT_PASSWORD_TEXT), ' ', index.h("amplify-button", { variant: "anchor", onClick: () => this.handleAuthStateChange(authTypes.AuthState.ForgotPassword), "data-test": "sign-in-forgot-password-link" }, core.I18n.get(Translations.Translations.RESET_PASSWORD_TEXT)))),
            required: true,
            handleInputChange: this.handleFormFieldInputChange('password'),
            inputProps: {
                'data-test': 'sign-in-password-input',
            },
        });
        this.newFormFields = [...formFieldInputs];
    }
    buildFormFields() {
        if (this.formFields.length === 0) {
            this.buildDefaultFormFields();
        }
        else {
            const newFields = [];
            this.formFields.forEach(field => {
                const newField = Object.assign({}, field);
                // TODO: handle hint better
                if (newField.type === 'password') {
                    newField['hint'] = helpers.isHintValid(newField) ? (index.h("div", null, core.I18n.get(Translations.Translations.FORGOT_PASSWORD_TEXT), ' ', index.h("amplify-button", { variant: "anchor", onClick: () => this.handleAuthStateChange(authTypes.AuthState.ForgotPassword), "data-test": "sign-in-forgot-password-link" }, core.I18n.get(Translations.Translations.RESET_PASSWORD_TEXT)))) : (newField['hint']);
                }
                newField['handleInputChange'] = event => this.handleFormFieldInputWithCallback(event, field);
                this.setFieldValue(newField, this.signInAttributes);
                newFields.push(newField);
            });
            this.newFormFields = newFields;
        }
    }
    setFieldValue(field, formAttributes) {
        switch (field.type) {
            case 'username':
            case 'email':
                if (field.value === undefined) {
                    formAttributes.userInput = '';
                }
                else {
                    formAttributes.userInput = field.value;
                }
                break;
            case 'phone_number':
                if (field.dialCode) {
                    this.phoneNumber.countryDialCodeValue = field.dialCode;
                }
                this.phoneNumber.phoneNumberValue = field.value;
                break;
            case 'password':
                if (field.value === undefined) {
                    formAttributes.password = '';
                }
                else {
                    formAttributes.password = field.value;
                }
                break;
        }
    }
    render() {
        return (index.h(index.Host, null, index.h("amplify-form-section", { headerText: core.I18n.get(this.headerText), handleSubmit: this.handleSubmit, testDataPrefix: 'sign-in' }, index.h("div", { slot: "subtitle" }, index.h("slot", { name: "header-subtitle" })), index.h("slot", { name: "federated-buttons" }, index.h("amplify-federated-buttons", { handleAuthStateChange: this.handleAuthStateChange, federated: this.federated })), !core.isEmpty(this.federated) && index.h("amplify-strike", null, "or"), index.h("amplify-auth-fields", { formFields: this.newFormFields }), index.h("div", { slot: "amplify-form-section-footer", class: "sign-in-form-footer" }, index.h("slot", { name: "footer" }, !this.hideSignUp && (index.h("slot", { name: "secondary-footer-content" }, index.h("span", null, core.I18n.get(Translations.Translations.NO_ACCOUNT_TEXT), ' ', index.h("amplify-button", { variant: "anchor", onClick: () => this.handleAuthStateChange(authTypes.AuthState.SignUp), "data-test": "sign-in-create-account-link" }, core.I18n.get(Translations.Translations.CREATE_ACCOUNT_TEXT))))), index.h("div", { class: this.hideSignUp ? 'full-width-footer-content' : '' }, index.h("slot", { name: "primary-footer-content" }, index.h("amplify-button", { type: "submit", disabled: this.loading, "data-test": "sign-in-sign-in-button" }, this.loading ? (index.h("amplify-loading-spinner", null)) : (index.h("span", null, core.I18n.get(this.submitButtonText)))))))))));
    }
    static get watchers() { return {
        "formFields": ["formFieldsHandler"]
    }; }
};
AmplifySignIn.style = amplifySignInCss;

const amplifySignUpCss = "amplify-sign-up{--footer-font-family:var(--amplify-font-family);--footer-font-size:var(--amplify-text-sm);--footer-color:var(--amplify-grey);--font-weight:var(--amplify-font-weight)}.sign-up-form-footer{font-family:var(--footer-font-family);font-size:var(--footer-font-size);color:var(--footer-color);font-weight:--font-weight;display:-ms-flexbox;display:flex;-ms-flex-direction:column-reverse;flex-direction:column-reverse;-ms-flex-align:center;align-items:center;-ms-flex-pack:distribute;justify-content:space-around}.sign-up-form-footer amplify-button{margin-bottom:0.625rem}@media (min-width: 672px){.sign-up-form-footer{display:-ms-flexbox;display:flex;-ms-flex-direction:row;flex-direction:row;-ms-flex-pack:justify;justify-content:space-between;-ms-flex-align:baseline;align-items:baseline}.sign-up-form-footer amplify-button{margin-bottom:0}}.sign-up-form-footer *+*{margin-bottom:15px}";

const AmplifySignUp = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /** Fires when sign up form is submitted */
        this.handleSubmit = event => this.signUp(event);
        /** Override for handling the Auth.SignUp API call */
        this.handleSignUp = params => this.authSignUp(params);
        /** Used for header text in sign up component */
        this.headerText = Translations.Translations.SIGN_UP_HEADER_TEXT;
        /** Used for the submit button text in sign up component */
        this.submitButtonText = Translations.Translations.SIGN_UP_SUBMIT_BUTTON_TEXT;
        /** Used for the submit button text in sign up component */
        this.haveAccountText = Translations.Translations.SIGN_UP_HAVE_ACCOUNT_TEXT;
        /** Text used for the sign in hyperlink */
        this.signInText = Translations.Translations.SIGN_IN_TEXT;
        /**
         * Form fields allows you to utilize our pre-built components such as username field, code field, password field, email field, etc.
         * by passing an array of strings that you would like the order of the form to be in. If you need more customization, such as changing
         * text for a label or adjust a placeholder, you can follow the structure below in order to do just that.
         * ```
         * [
         *  {
         *    type: string,
         *    label: string,
         *    placeholder: string,
         *    hint: string | Functional Component | null,
         *    required: boolean
         *  }
         * ]
         * ```
         */
        this.formFields = [];
        /** Auth state change handler for this component
         * e.g. SignIn -> 'Create Account' link -> SignUp
         */
        // prettier-ignore
        this.handleAuthStateChange = helpers.dispatchAuthStateChangeEvent;
        /** Username Alias is used to setup authentication with `username`, `email` or `phone_number`  */
        this.usernameAlias = 'username';
        // private userInput: string | PhoneNumberInterface;
        this.newFormFields = [];
        this.phoneNumber = {
            countryDialCodeValue: constants.COUNTRY_DIAL_CODE_DEFAULT,
            phoneNumberValue: null,
        };
        this.loading = false;
        this.signUpAttributes = {
            username: '',
            password: '',
            attributes: {},
        };
    }
    handleFormFieldInputChange(fieldType) {
        switch (fieldType) {
            case 'username':
                return event => (this.signUpAttributes.username = event.target.value);
            case 'password':
                return event => (this.signUpAttributes.password = event.target.value);
            case 'email':
                return event => (this.signUpAttributes.attributes.email = event.target.value);
            case 'phone_number':
                return event => helpers.handlePhoneNumberChange(event, this.phoneNumber);
            default:
                return event => (this.signUpAttributes.attributes[fieldType] = event.target.value);
        }
    }
    handleFormFieldInputWithCallback(event, field) {
        const fnToCall = field['handleInputChange']
            ? field['handleInputChange']
            : (event, cb) => {
                cb(event);
            };
        const callback = this.handleFormFieldInputChange(field.type);
        fnToCall(event, callback.bind(this));
    }
    async authSignUp(params) {
        const data = await Auth.Auth.signUp(params);
        if (!data) {
            throw new Error(Translations.Translations.SIGN_UP_FAILED);
        }
        return data;
    }
    assignPhoneNumberToSignUpAttributes() {
        if (this.phoneNumber.phoneNumberValue) {
            try {
                this.signUpAttributes.attributes.phone_number = helpers.composePhoneNumberInput(this.phoneNumber);
            }
            catch (error) {
                helpers.dispatchToastHubEvent(error);
            }
        }
    }
    assignUserNameAliasToSignUpAttributes() {
        switch (this.usernameAlias) {
            case 'email':
            case 'phone_number':
                this.signUpAttributes.username = this.signUpAttributes.attributes[this.usernameAlias];
                break;
        }
    }
    assignFormInputToSignUpAttributes() {
        this.assignPhoneNumberToSignUpAttributes();
        this.assignUserNameAliasToSignUpAttributes();
    }
    validateSignUpAttributes() {
        if (!this.signUpAttributes.username) {
            if (this.usernameAlias === authTypes.UsernameAlias.email) {
                throw new Error(Translations.Translations.EMPTY_EMAIL);
            }
            else if (this.usernameAlias === authTypes.UsernameAlias.phone_number) {
                throw new Error(Translations.Translations.EMPTY_PHONE);
            }
            else {
                throw new Error(Translations.Translations.EMPTY_USERNAME);
            }
        }
        if (this.signUpAttributes.username.indexOf(' ') >= 0) {
            throw new Error(Translations.Translations.USERNAME_REMOVE_WHITESPACE);
        }
        if (this.signUpAttributes.password !== this.signUpAttributes.password.trim()) {
            throw new Error(Translations.Translations.PASSWORD_REMOVE_WHITESPACE);
        }
    }
    // TODO: Add validation
    // TODO: Prefix
    async signUp(event) {
        if (event) {
            event.preventDefault();
        }
        if (!Auth.Auth || typeof Auth.Auth.signUp !== 'function') {
            throw new Error(constants.NO_AUTH_MODULE_FOUND);
        }
        this.loading = true;
        this.assignFormInputToSignUpAttributes();
        try {
            this.validateSignUpAttributes();
            const data = await this.handleSignUp(this.signUpAttributes);
            if (data.userConfirmed) {
                await authHelpers.handleSignIn(this.signUpAttributes.username, this.signUpAttributes.password, this.handleAuthStateChange);
            }
            else {
                const signUpAttrs = Object.assign({}, this.signUpAttributes);
                this.handleAuthStateChange(authTypes.AuthState.ConfirmSignUp, Object.assign(Object.assign({}, data.user), { signUpAttrs }));
            }
        }
        catch (error) {
            helpers.dispatchToastHubEvent(error);
        }
        finally {
            this.loading = false;
        }
    }
    buildDefaultFormFields() {
        switch (this.usernameAlias) {
            case 'email':
                this.newFormFields = [
                    {
                        type: 'email',
                        placeholder: core.I18n.get(Translations.Translations.SIGN_UP_EMAIL_PLACEHOLDER),
                        required: true,
                        handleInputChange: this.handleFormFieldInputChange('email'),
                        inputProps: {
                            'data-test': 'sign-up-email-input',
                            autocomplete: 'username',
                        },
                    },
                    {
                        type: 'password',
                        placeholder: core.I18n.get(Translations.Translations.SIGN_UP_PASSWORD_PLACEHOLDER),
                        required: true,
                        handleInputChange: this.handleFormFieldInputChange('password'),
                        inputProps: {
                            'data-test': 'sign-up-password-input',
                            autocomplete: 'new-password',
                        },
                    },
                    {
                        type: 'phone_number',
                        required: true,
                        handleInputChange: this.handleFormFieldInputChange('phone_number'),
                        inputProps: {
                            'data-test': 'sign-up-phone-number-input',
                            autocomplete: 'tel-national',
                        },
                    },
                ];
                break;
            case 'phone_number':
                this.newFormFields = [
                    {
                        type: 'phone_number',
                        required: true,
                        handleInputChange: this.handleFormFieldInputChange('phone_number'),
                        inputProps: {
                            'data-test': 'sign-up-phone-number-input',
                            autocomplete: 'username',
                        },
                    },
                    {
                        type: 'password',
                        placeholder: core.I18n.get(Translations.Translations.SIGN_UP_PASSWORD_PLACEHOLDER),
                        required: true,
                        handleInputChange: this.handleFormFieldInputChange('password'),
                        inputProps: {
                            'data-test': 'sign-up-password-input',
                            autocomplete: 'new-password',
                        },
                    },
                    {
                        type: 'email',
                        placeholder: core.I18n.get(Translations.Translations.SIGN_UP_EMAIL_PLACEHOLDER),
                        required: true,
                        handleInputChange: this.handleFormFieldInputChange('email'),
                        inputProps: {
                            'data-test': 'sign-up-email-input',
                            autocomplete: 'email',
                        },
                    },
                ];
                break;
            case 'username':
            default:
                this.newFormFields = [
                    {
                        type: 'username',
                        placeholder: core.I18n.get(Translations.Translations.SIGN_UP_USERNAME_PLACEHOLDER),
                        required: true,
                        handleInputChange: this.handleFormFieldInputChange('username'),
                        inputProps: {
                            'data-test': 'sign-up-username-input',
                            autocomplete: 'username',
                        },
                    },
                    {
                        type: 'password',
                        placeholder: core.I18n.get(Translations.Translations.SIGN_UP_PASSWORD_PLACEHOLDER),
                        required: true,
                        handleInputChange: this.handleFormFieldInputChange('password'),
                        inputProps: {
                            'data-test': 'sign-up-password-input',
                            autocomplete: 'new-password',
                        },
                    },
                    {
                        type: 'email',
                        placeholder: core.I18n.get(Translations.Translations.SIGN_UP_EMAIL_PLACEHOLDER),
                        required: true,
                        handleInputChange: this.handleFormFieldInputChange('email'),
                        inputProps: {
                            'data-test': 'sign-up-email-input',
                        },
                    },
                    {
                        type: 'phone_number',
                        required: true,
                        handleInputChange: this.handleFormFieldInputChange('phone_number'),
                        inputProps: {
                            'data-test': 'sign-up-phone-number-input',
                        },
                    },
                ];
                break;
        }
    }
    buildFormFields() {
        if (this.formFields.length === 0) {
            this.buildDefaultFormFields();
        }
        else {
            const newFields = [];
            this.formFields.forEach(field => {
                const newField = Object.assign({}, field);
                newField['handleInputChange'] = event => this.handleFormFieldInputWithCallback(event, field);
                this.setFieldValue(field, this.signUpAttributes);
                newFields.push(newField);
            });
            this.newFormFields = newFields;
        }
    }
    setFieldValue(field, formAttributes) {
        switch (field.type) {
            case 'username':
                if (field.value === undefined) {
                    formAttributes.username = '';
                }
                else {
                    formAttributes.username = field.value;
                }
                break;
            case 'password':
                if (field.value === undefined) {
                    formAttributes.password = '';
                }
                else {
                    formAttributes.password = field.value;
                }
                break;
            case 'email':
                formAttributes.attributes.email = field.value;
                break;
            case 'phone_number':
                if (field.dialCode) {
                    this.phoneNumber.countryDialCodeValue = field.dialCode;
                }
                this.phoneNumber.phoneNumberValue = field.value;
                break;
            default:
                formAttributes.attributes[field.type] = field.value;
                break;
        }
    }
    componentWillLoad() {
        helpers.checkUsernameAlias(this.usernameAlias);
        this.buildFormFields();
    }
    formFieldsHandler() {
        this.buildFormFields();
    }
    render() {
        return (index.h(index.Host, null, index.h("amplify-form-section", { headerText: core.I18n.get(this.headerText), handleSubmit: this.handleSubmit, testDataPrefix: 'sign-up' }, index.h("div", { slot: "subtitle" }, index.h("slot", { name: "header-subtitle" })), index.h("amplify-auth-fields", { formFields: this.newFormFields }), index.h("div", { class: "sign-up-form-footer", slot: "amplify-form-section-footer" }, index.h("slot", { name: "footer" }, index.h("slot", { name: "secondary-footer-content" }, index.h("span", null, core.I18n.get(this.haveAccountText), ' ', index.h("amplify-button", { variant: "anchor", onClick: () => this.handleAuthStateChange(authTypes.AuthState.SignIn), "data-test": "sign-up-sign-in-link" }, core.I18n.get(this.signInText)))), index.h("slot", { name: "primary-footer-content" }, index.h("amplify-button", { type: "submit", "data-test": "sign-up-create-account-button", disabled: this.loading }, this.loading ? (index.h("amplify-loading-spinner", null)) : (index.h("span", null, core.I18n.get(this.submitButtonText))))))))));
    }
    static get watchers() { return {
        "formFields": ["formFieldsHandler"]
    }; }
};
AmplifySignUp.style = amplifySignUpCss;

const logger$2 = new core.Logger('AmplifyVerifyContact');
const AmplifyVerifyContact = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /** Authentication state handler */
        this.handleAuthStateChange = helpers.dispatchAuthStateChangeEvent;
        this.loading = false;
    }
    handleSubmit(event) {
        event.preventDefault();
        this.verifyAttr ? this.submit(this.code) : this.verify(this.contact);
    }
    async submit(code) {
        const attr = this.verifyAttr;
        if (!Auth.Auth || typeof Auth.Auth.verifyCurrentUserAttributeSubmit !== 'function') {
            throw new Error(constants.NO_AUTH_MODULE_FOUND);
        }
        this.loading = true;
        try {
            const data = await Auth.Auth.verifyCurrentUserAttributeSubmit(attr, code);
            logger$2.debug(data);
            this.handleAuthStateChange(authTypes.AuthState.SignedIn, this.user);
            this.verifyAttr = null;
        }
        catch (error) {
            helpers.dispatchToastHubEvent(error);
            logger$2.error(error);
        }
        finally {
            this.loading = false;
        }
    }
    async verify(contact) {
        if (!contact) {
            logger$2.error('Neither Email nor Phone Number selected');
            return;
        }
        if (!Auth.Auth || typeof Auth.Auth.verifyCurrentUserAttribute !== 'function') {
            throw new Error(constants.NO_AUTH_MODULE_FOUND);
        }
        this.loading = true;
        try {
            const data = await Auth.Auth.verifyCurrentUserAttribute(contact);
            logger$2.debug(data);
            this.verifyAttr = contact;
        }
        catch (error) {
            helpers.dispatchToastHubEvent(error);
            logger$2.error(error);
        }
        finally {
            this.loading = false;
        }
    }
    handleInputChange(event) {
        const inputName = event.target.name;
        if (inputName === 'code') {
            this.code = event.target.value;
        }
        else if (inputName === 'contact') {
            this.contact = event.target.value;
        }
    }
    renderSubmit() {
        return (index.h("div", null, index.h("amplify-input", { inputProps: {
                autocomplete: 'off',
                'data-test': 'verify-contact-code-input',
            }, name: "code", placeholder: core.I18n.get(Translations.Translations.CODE_PLACEHOLDER), handleInputChange: event => this.handleInputChange(event) })));
    }
    renderVerify() {
        const user = this.user;
        if (!user) {
            logger$2.debug('No user to verify');
            return null;
        }
        const { unverified } = user;
        if (!unverified) {
            logger$2.debug('Unverified variable does not exist on user');
            return null;
        }
        const { email, phone_number } = unverified;
        return (index.h("div", null, email && (index.h("amplify-radio-button", { label: core.I18n.get(Translations.Translations.VERIFY_CONTACT_EMAIL_LABEL), key: "email", name: "contact", value: "email", handleInputChange: event => this.handleInputChange(event), inputProps: {
                'data-test': 'verify-contact-email-radio',
            } })), phone_number && (index.h("amplify-radio-button", { label: core.I18n.get(Translations.Translations.VERIFY_CONTACT_PHONE_LABEL), key: "phone_number", name: "contact", value: "phone_number", handleInputChange: event => this.handleInputChange(event), inputProps: {
                'data-test': 'verify-contact-phone-number-radio',
            } }))));
    }
    render() {
        return (index.h(index.Host, null, index.h("amplify-form-section", { handleSubmit: event => this.handleSubmit(event), headerText: core.I18n.get(Translations.Translations.VERIFY_CONTACT_HEADER_TEXT), loading: this.loading, secondaryFooterContent: index.h("span", null, index.h("amplify-button", { variant: "anchor", onClick: () => this.handleAuthStateChange(authTypes.AuthState.SignedIn, this.user), "data-test": "verify-contact-skip-link" }, "Skip")), submitButtonText: this.verifyAttr
                ? core.I18n.get(Translations.Translations.VERIFY_CONTACT_SUBMIT_LABEL)
                : core.I18n.get(Translations.Translations.VERIFY_CONTACT_VERIFY_LABEL) }, this.verifyAttr ? this.renderSubmit() : this.renderVerify())));
    }
};

exports.amplify_confirm_sign_in = AmplifyConfirmSignIn;
exports.amplify_confirm_sign_up = AmplifyConfirmSignUp;
exports.amplify_forgot_password = AmplifyForgotPassword;
exports.amplify_require_new_password = AmplifyRequireNewPassword;
exports.amplify_sign_in = AmplifySignIn;
exports.amplify_sign_up = AmplifySignUp;
exports.amplify_verify_contact = AmplifyVerifyContact;
